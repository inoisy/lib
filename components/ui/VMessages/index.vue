<script>

// Mixins
import Colorable from '../../../mixins/colorable';
import Themeable from '../../../mixins/themeable';

import mixins from '../../../utils/mixins';

// Utilities
import { getSlot } from '../../../utils/helpers';


export default mixins(Colorable, Themeable).extend({
    name: 'VMessages',

    props: {
        value: {
            type: Array,
            default: () => [],
        },
    },

    methods: {
        genChildren() {
            return this.$createElement('transition-group', {
                staticClass: 'v-messages__wrapper',
                attrs: {
                    name: 'message-transition',
                    tag: 'div',
                },
            }, this.value.map(this.genMessage));
        },
        genMessage(message, key) {
            return this.$createElement('div', {
                staticClass: 'v-messages__message',
                key: key,
            }, getSlot(this, 'default', { message, key }) || [message]);
        },
    },

    render(h) {
        return h('div', this.setTextColor(this.color, {
            staticClass: 'v-messages',
            class: this.themeClasses,
        }), [this.genChildren()]);
    },
});

</script>
<style lang="scss">


    $messages-font-size: 12px !default;
    $messages-line-height: $messages-font-size !default;
    $messages-min-height: 14px !default;
    // $messages-font-size: 12px;
    $messages-min-height: unset;

    /* Theme */
    @include theme(v-messages) using ($material) {
        color: map-deep-get($material, 'text', 'secondary');
    }

    .v-messages {
        position: relative;
        flex: 1 1 auto;
        min-width: 1px; // Ensure takes up space with no messages and inside of flex
        min-height: $messages-min-height;

        //   +ltr()
        text-align: left;
        font-size: $messages-font-size;

        //   +rtl()
        //     text-align: right

        &__message {
            word-wrap: break-word;
            line-height: $messages-line-height;
            word-break: break-word;
            overflow-wrap: break-word;
            hyphens: auto;
        }
    }

</style>
