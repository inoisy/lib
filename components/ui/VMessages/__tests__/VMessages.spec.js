// Components
import VMessages from '../index';

// Utilities
import {
    mount,
} from '@vue/test-utils';

describe('VMessages.js', () => {
    let mountFunction;

    beforeEach(() => {
        mountFunction = (options = {}) => mount(VMessages, {
            ...options,
        });
    });

    it('should have a default array', () => {
        const wrapper = mountFunction();

        expect(Array.isArray(wrapper.vm.value)).toBe(true);
    });

    it('should show messages', async () => {
        const wrapper = mountFunction({
            propsData: {
                value: ['foo', 'bar'],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();

        await wrapper.setProps({ value: [] });
        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should allow HTML', () => {
        const wrapper = mountFunction({
            propsData: {
                value: ['<a href="#">a link</a>'],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not allow HTML', () => {
        const wrapper = mount(VMessages, {
            propsData: {
                value: ['<a href="#">a link</a>'],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should accept a scoped slot', () => {
        const wrapper = mount(VMessages, {
            propsData: { value: ['Foo'] },
            scopedSlots: {
                default(props) {
                    return this.$createElement('div', props.message);
                },
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});
