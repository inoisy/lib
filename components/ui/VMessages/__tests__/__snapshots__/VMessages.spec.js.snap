// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VMessages.js should accept a scoped slot 1`] = `
<div class="v-messages theme--light">
  <transition-group-stub tag="div"
                         name="message-transition"
                         class="v-messages__wrapper"
  >
    <div class="v-messages__message">
      <div>
        Foo
      </div>
    </div>
  </transition-group-stub>
</div>
`;

exports[`VMessages.js should allow HTML 1`] = `
<div class="v-messages theme--light">
  <transition-group-stub tag="div"
                         name="message-transition"
                         class="v-messages__wrapper"
  >
    <div class="v-messages__message">
      &lt;a href="#"&gt;a link&lt;/a&gt;
    </div>
  </transition-group-stub>
</div>
`;

exports[`VMessages.js should not allow HTML 1`] = `
<div class="v-messages theme--light">
  <transition-group-stub tag="div"
                         name="message-transition"
                         class="v-messages__wrapper"
  >
    <div class="v-messages__message">
      &lt;a href="#"&gt;a link&lt;/a&gt;
    </div>
  </transition-group-stub>
</div>
`;

exports[`VMessages.js should show messages 1`] = `
<div class="v-messages theme--light">
  <transition-group-stub tag="div"
                         name="message-transition"
                         class="v-messages__wrapper"
  >
    <div class="v-messages__message">
      foo
    </div>
    <div class="v-messages__message">
      bar
    </div>
  </transition-group-stub>
</div>
`;

exports[`VMessages.js should show messages 2`] = `
<div class="v-messages theme--light">
  <transition-group-stub tag="div"
                         name="message-transition"
                         class="v-messages__wrapper"
  >
  </transition-group-stub>
</div>
`;
